segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java

```diff
@@ -386,6 +386,13 @@ private void acknowledgeAlreadyProcessedOperation(StorageOperation operation) {
      */
     private void aggregateAppendOperation(StorageOperation operation, AggregatedAppendOperation aggregatedAppend, int maxLength) {
         long remainingLength = operation.getLength();
+        if (operation.getStreamSegmentOffset() < aggregatedAppend.getLastStreamSegmentOffset()) {
+            // The given operation begins before the AggregatedAppendOperation. This is likely due to it having been
+            // partially written to Storage prior to some recovery event. We must make sure we only include the part that
+            // has not yet been written.
+            remainingLength -= aggregatedAppend.getLastStreamSegmentOffset() - operation.getStreamSegmentOffset();
```

Comment:
Let's add a debug log message to indicate that this has happened.

Reply:
Fixed.